File: day09/day09.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# read input
cipher = []

fp = open('input', 'r')
for line in fp:
    cipher.append(line.strip())
fp.close()

def xmas(preamble_size = 25):
    for idx in range(preamble_size, len(cipher)):
        calc = []
        for x in range(idx - preamble_size, idx):
            for y in range(x + 1, idx):
                calc.append(int(cipher[x]) + int(cipher[y]))
        if int(cipher[idx]) not in calc:
            return int(cipher[idx])

def xmas2(target):
    for idx1 in range(0, len(cipher)):
        numbers = []
        total = 0

        val1 = int(cipher[idx1])
        numbers.append(val1)
        total += val1

        for idx2 in range(idx1 + 1, len(cipher)):
            val2 = int(cipher[idx2])
            numbers.append(val2)
            total += val2

            if total == target:
                numbers.sort()
                return numbers[0] + numbers[len(numbers) - 1]

# part 1
part1 = xmas(25)
print("part1", part1)

part2 = xmas2(part1)
print("part2", part2)

File: advent2020/src/day09.rs
#![feature(test)]

extern crate test;

use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

fn xmas(cipher: &Vec<u64>, size: u64) -> u64 {
    for idx in size..cipher.len() as u64 {
        let mut calc: Vec<u64> = Vec::new();
        for x in idx - size..idx {
            for y in x + 1..idx {
                let value = cipher[x as usize] + cipher[y as usize];
                calc.push(value);
            }
        }
        if calc.contains(&cipher[idx as usize]) == false {
            return cipher[idx as usize];
        }
    }
    return 0;
}

fn xmas2(cipher: &Vec<u64>, target: u64) -> u64 {
    let cipherlen = cipher.len() as u64;
    for idx1 in 0..cipherlen as u64 {
        let mut numbers: Vec<u64> = Vec::new();
        let mut total: u64 = 0;

        let val1 = cipher[idx1 as usize];
        numbers.push(val1);
        total += val1;

        for idx2 in idx1 + 1..cipherlen as u64 {
            let val2 = cipher[idx2 as usize];
            numbers.push(val2);
            total += val2;

            if total == target {
                numbers.sort();
                return numbers[0] + numbers[numbers.len() - 1]
            }
        }
    }
    return 0;
}

fn main() {
    let mut cipher: Vec<u64> = Vec::new();
    let f = File::open("../day09/input").unwrap();
    let file = BufReader::new(&f);
    for line in file.lines() {
        cipher.push(line.unwrap().parse::<u64>().unwrap());
    }

    let part1 = xmas(&cipher, 25);
    println!("part1 {}", part1);
    let part2 = xmas2(&cipher, part1);
    println!("part2 {}", part2);
}

#[bench]
fn bench_day09(b: &mut test::Bencher) {
    b.iter(|| main());
}
